### PR TITLE
o/devicestate: introduce DeviceManager.Unregister

### PR DESCRIPTION
**Pull request by pedronis**
this makes snapd forget the device serial and key

and there's an option to block re-registration until the next reboot

for now this allowed only on classic for generic or canonical models
(these models get their serials from the store)

TODO:
 - add API access to this
 - support actually deleting the device key, this requires asserts
   package changes

### CODE REVIEW

File: overlord/devicestate/devicemgr.go [hunk 1]
```diff
@@ -1360,6 +1372,44 @@ func (m *DeviceManager) Registered() <-chan struct{} {
...
+	if opts == nil {
+		opts = &UnregisterOptions{}
+	}
+	if opts.NoRegistrationUntilReboot {
+		if err := os.MkdirAll(dirs.SnapRunDir, 0755); err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(filepath.Join(dirs.SnapRunDir, "noregister"), nil, 0644); err != nil {
+			return err
+		}
```
Comment (mardy) [inline overlord/devicestate/devicemgr.go:1399]: If there's no space left on the device and the sentinel file cannot be created, would it still be beneficial to unregister the device?

Maybe we still want to return an error in that case, but move the creation of the sentinel at the end of this function instead?

File: overlord/devicestate/devicestate_serial_test.go [hunk 2]
```diff
@@ -2017,3 +2021,181 @@ func (s *deviceMgrSerialSuite) TestFullDeviceRegistrationUC20Happy(c *C) {
...
+
+	privKey, err = devicestate.KeypairManager(s.mgr).Get(serial.DeviceKey().ID())
+	c.Assert(err, IsNil)
+	c.Check(privKey, NotNil)
+	c.Check(device.KeyID, Equals, privKey.PublicKey().ID())
+	// different from previous key
+	c.Check(device.KeyID, Not(Equals), keyID1)
+}
+
+func (s *deviceMgrSerialSuite) TestFullDeviceRegistrationBlockedByNoResister(c *C) {
```
Comment (mardy) [inline overlord/devicestate/devicestate_serial_test.go:2155]: resister -> register

File: overlord/devicestate/devicemgr.go [hunk 3]
```diff
@@ -489,6 +495,12 @@ func (m *DeviceManager) ensureOperational() error {
 		}
 	}
 
+	// registration is blocked until reboot
+	if osutil.FileExists(filepath.Join(dirs.SnapRunDir, "noregister")) {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/devicestate/devicemgr.go:499]: I agree, the other checks seem to be only checking the state to return early, could this be moved up close to `if m.noRegister` check?

File: overlord/devicestate/devicemgr.go [hunk 4]
```diff
@@ -1360,6 +1372,44 @@ func (m *DeviceManager) Registered() <-chan struct{} {
...
+			return err
+		}
+	}
+	device.Serial = ""
+	device.KeyID = ""
+	device.SessionMacaroon = ""
+	if err := m.setDevice(device); err != nil {
+		return err
+	}
+	// TODO: delete key
```
Comment (stolowski) [inline overlord/devicestate/devicemgr.go:1407]: What key is it about, can you expand this comment?

**Review by stolowski — COMMENTED** [review]
Looks good, a few nitpicks

File: overlord/devicestate/devicemgr.go (hunk 4 above)
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go:1407]: Device key(pair), but yeah maybe it's better to make it clear and expand the comment a little

File: overlord/devicestate/devicemgr.go [hunk 5]
```diff
@@ -1360,6 +1372,44 @@ func (m *DeviceManager) Registered() <-chan struct{} {
 	return m.reg
 }
 
+type UnregisterOptions struct {
+	NoRegistrationUntilReboot bool
+}
+
+// Unregister unregisters the device forgetting its serial
+// plus the additional behavior described by the UnregisterOptions
+func (m *DeviceManager) Unregister(opts *UnregisterOptions) error {
```
Comment (bboozzoo) [inline overlord/devicestate/devicemgr.go:1381]: This is called from contexts where the caller already has the state lock?

**Review by bboozzoo — APPROVED** [review]
LGTM

File: overlord/devicestate/devicemgr.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/devicestate/devicemgr.go:499]: I added a comment,  the reason is here, is to do mostly in-memory checks before doing this filesystem check

File: overlord/devicestate/devicemgr.go (hunk 5 above)
Comment (pedronis) [reply overlord/devicestate/devicemgr.go:1381]: yes

File: overlord/devicestate/devicemgr.go (hunk 1 above)
Comment (pedronis) [reply overlord/devicestate/devicemgr.go:1402]: the use cases we have would be in controlled environments, so we want to error

**Comment by mvo5** [comment]
Adding milestone because the store would really like to have this for GH to be able to unregister cloned images.

**Comment by pedronis** [comment]
> Adding milestone because the store would really like to have this for GH to be able to unregister cloned images.

@mvo5 this needs at least API work as well though

**Pull request merged**